pkg/extensions/tbtc/tbtc.go

```diff
@@ -839,7 +813,39 @@ func (t *tbtc) shouldMonitorDeposit(depositAddress string) bool {
 			depositAddress,
 			err,
 		)
-		return false // return false but don't cache the result in case of error
+		// return false but don't cache the result in case of an error;
+		// it gives a chance to retry the monitoring later but sooner
+		// than notMonitoredDepositsCache caching period.
+		return false
+	}
+
+	hasInitialState, err := utils.ConfirmWithTimeoutDefaultBackoff(
```

Comment:
I performed several tests and I think we have a problem here. Suppose a test scenario:

1. A deposit is opened and clients start `retrieve pubkey` monitoring. All of them cache this deposit in their `monitoredDepositsCache`.
2. The public key is retrieved by a third-party. All clients stop their `retrieve pubkey` monitoring instances.
3. Before `24 hours` pass, clients lose their ETH connections for a while. After restoring, past events are checked. This causes an occurrence of a duplicated `Created` event telling about the deposit creation once again.
4. Clients try to start a new instance of the `retrieve pubkey` monitoring for the same deposit. Because they all have that deposit in their `monitoredDepositsCache`, they all decide to start a new monitoring instance without checking the actual deposit state.

Reply:
Good catch, addressed in 77c355b. I will quote here explanation from the commit message:


The previous approach assumed that if the given deposit address is cached in `monitoredDepositsCache`, `shouldMonitorDeposit` returns `true` without executing any further checks.

Together with `acquireMonitoringLock` and `releaseMonitoringLock`, it worked fine for most of the cases but not for all of them. Specifically, for the case when the deposit has already transitioned to the next state, we could run into a situation when the monitoring lock has already been released and `shouldMonitorDeposit` returns `true` just because the client has a member in that deposit, without executing the initial state check.

The order of cache checks has been changed to address it.

Also, renamed `monitoredDepositsCache` to `memberDepositsCache`, and `notMonitoredDepositsCache` to `notMemberDepositsCache` to state it clearly those caches hold information whether the client has or has not a member in the given deposit.

The new approach is:
1. Check if the given deposit is in `notMonitoredDepositsCache`. If so, we are sure the client is not interested in monitoring this deposit no matter what - it has no member in it.

2. Check the initial state of the deposit to do not start monitoring for duplicate events from the past. If the initial state is not as expected, the client is not interested in monitoring this deposit as the event triggering the monitoring is something from the past that does not match the current chain state.

3. Check if the given deposit is in `memberDepositCache`. If so, start the monitoring. This check allows to do not repeat `getSignerIndex` call for the same deposit more than one time. If the deposit is not in `memberDepositCache`, execute `getSignerIndex` call, check members, update caches and return `true` or `false` depending on if the client has a member in the deposit.

Reply:
Ok, that change looks good. Tested and the problem has gone.